### ISSUE TITLE
System.ArgumentNullException on Update-Database command (Access97 Database, FirstMigration)

### ISSUE THREAD

**Issue by juliokele** [bubibubi/EntityFrameworkCore.Jet#106]
i have System.ArgumentNullException on Update-Database command (Access97 Database, FirstMigration):

```
PM> Add-Migration FirstMigration
Build started...
Build succeeded.
To undo this action, use Remove-Migration.

PM> Update-Database -v
Using project 'Access97EF.ConsoleUI'.
Using startup project 'Access97EF.ConsoleUI'.
Build started...
Build succeeded.
C:\Users\user\.nuget\packages\microsoft.entityframeworkcore.tools\3.1.20\tools\net461\win-x86\ef.exe database update --verbose --no-color --prefix-output --assembly c:\Temp\Access97EF\Access97EF.Console\bin\Debug\net48\Access97EF.ConsoleUI.exe --startup-assembly c:\Temp\Access97EF\Access97EF.Console\bin\Debug\net48\Access97EF.ConsoleUI.exe --project-dir c:\Temp\Access97EF\Access97EF.Console\ --language C# --working-dir C:\Temp\Access97EF --root-namespace Access97EF.ConsoleUI
Using assembly 'Access97EF.ConsoleUI'.
Using startup assembly 'Access97EF.ConsoleUI'.
Using application base 'c:\Temp\Access97EF\Access97EF.Console\bin\Debug\net48'.
Using working directory 'c:\Temp\Access97EF\Access97EF.Console'.
Using root namespace 'Access97EF.ConsoleUI'.
Using project directory 'c:\Temp\Access97EF\Access97EF.Console\'.
Using configuration file 'c:\Temp\Access97EF\Access97EF.Console\bin\Debug\net48\Access97EF.ConsoleUI.exe.config'.
Using assembly 'Access97EF.ConsoleUI'.
Using startup assembly 'Access97EF.ConsoleUI'.
Using application base 'c:\Temp\Access97EF\Access97EF.Console\bin\Debug\net48'.
Using working directory 'c:\Temp\Access97EF\Access97EF.Console'.
Using root namespace 'Access97EF.ConsoleUI'.
Using project directory 'c:\Temp\Access97EF\Access97EF.Console\'.
Finding DbContext classes...
Finding IDesignTimeDbContextFactory implementations...
Finding application service provider...
Finding Microsoft.Extensions.Hosting service provider...
No static method 'CreateHostBuilder(string[])' was found on class 'Program'.
No application service provider was found.
Finding DbContext classes in the project...
Found DbContext 'ModelContext'.
Using context 'ModelContext'.
Finding design-time services for provider 'EntityFrameworkCore.Jet'...
Using design-time services from provider 'EntityFrameworkCore.Jet'.
Finding design-time services referenced by assembly 'Access97EF.ConsoleUI'.
No referenced design-time services were found.
Finding IDesignTimeServices implementations in assembly 'Access97EF.ConsoleUI'...
No design-time services were found.
Migrating using database '' on server 'c:\Temp\Access97EF\test.mdb'.
'ModelContext' disposed.
System.ArgumentNullException: Der Wert darf nicht NULL sein. -> The value cannot be NULL.
Parametername: input
   bei System.Text.RegularExpressions.Regex.Match(String input)
   bei EntityFrameworkCore.Jet.Data.JetStoreSchemaDefinition.JetStoreDatabaseHandling.ExtractFileNameFromConnectionString(String connectionString)
   bei EntityFrameworkCore.Jet.Data.JetConnection.get_DataSource()
   bei Microsoft.EntityFrameworkCore.Diagnostics.RelationalLoggerExtensions.LogConnectionOpening(IDiagnosticsLogger`1 diagnostics, IRelationalConnection connection, EventDefinition`2 definition)
   bei Microsoft.EntityFrameworkCore.Diagnostics.RelationalLoggerExtensions.ConnectionOpening(IDiagnosticsLogger`1 diagnostics, IRelationalConnection connection, DateTimeOffset startTime)
   bei Microsoft.EntityFrameworkCore.Storage.RelationalConnection.OpenDbConnection(Boolean errorsExpected)
   bei Microsoft.EntityFrameworkCore.Storage.RelationalConnection.Open(Boolean errorsExpected)
   bei Microsoft.EntityFrameworkCore.Migrations.Internal.MigrationCommandExecutor.ExecuteNonQuery(IEnumerable`1 migrationCommands, IRelationalConnection connection)
   bei EntityFrameworkCore.Jet.Storage.Internal.JetDatabaseCreator.Create()
   bei Microsoft.EntityFrameworkCore.Migrations.Internal.Migrator.Migrate(String targetMigration)
   bei Microsoft.EntityFrameworkCore.Design.Internal.MigrationsOperations.UpdateDatabase(String targetMigration, String contextType)
   bei Microsoft.EntityFrameworkCore.Design.OperationExecutor.UpdateDatabaseImpl(String targetMigration, String contextType)
   bei Microsoft.EntityFrameworkCore.Design.OperationExecutor.UpdateDatabase.<>c__DisplayClass0_0.<.ctor>b__0()
   bei Microsoft.EntityFrameworkCore.Design.OperationExecutor.OperationBase.Execute(Action action)
Der Wert darf nicht NULL sein. -> The value cannot be NULL.
Parametername: input
```
connection string: "Provider=Microsoft.ACE.OLEDB.12.0;Jet OLEDB:Engine Type=4;Data Source=c:\\Temp\\Access97EF\\test.mdb;"
EntityFrameworkCore.Jet: 3.1.0
Microsoft.EntityFrameworkCore.Tools: 3.1.20
Framework: net461

**Comment by ChrisJollyAU** [comment]
It looks like something is going wrong in the `ExtractFileNameFromConnectionString` function.

The likely candidate is a problem with your connection string. Are you sure you've set the connection string properly?

It looks like by the time it gets to performing a regex on it to find the value of the Data Source, your connection string is NULL

While it might not be part of the problem, I'm not too sure on using ACE.OLEDB.12.0 and setting the Engine Type to 4. Perhaps @lauxjpn can chime in on whether it would work

I've always thought using Microsoft.Jet.OLEDB.4.0 for the Provider ()like the below) is more recommended. I have a Access 95 db that I use as such
`Provider=Microsoft.Jet.OLEDB.4.0;Data Source=c:\\path\\db.mdb;`

**Comment by lcrumbling** [comment]
> I've always thought using Microsoft.Jet.OLEDB.4.0 for the Provider ()like the below) is more recommended. I have a Access 95 db that I use as such `Provider=Microsoft.Jet.OLEDB.4.0;Data Source=c:\\path\\db.mdb;`

FWIW, I believe 4.0 is Access 2000; OP mentions 97, which is 3.5:
Provider=Microsoft.Jet.OLEDB.3.5;Data Source=c:\\path\\db.mdb;

That being said, I believe 4.0 will work is 97. It -may- try to convert it to the 2000 format, though.

**Comment by juliokele** [comment]
> It looks like something is going wrong in the `ExtractFileNameFromConnectionString` function.
> 
> The likely candidate is a problem with your connection string. Are you sure you've set the connection string properly?
> 
> It looks like by the time it gets to performing a regex on it to find the value of the Data Source, your connection string is NULL
> 
> While it might not be part of the problem, I'm not too sure on using ACE.OLEDB.12.0 and setting the Engine Type to 4. Perhaps @lauxjpn can chime in on whether it would work
> 
> I've always thought using Microsoft.Jet.OLEDB.4.0 for the Provider ()like the below) is more recommended. I have a Access 95 db that I use as such `Provider=Microsoft.Jet.OLEDB.4.0;Data Source=c:\\path\\db.mdb;`

Scaffold on existing DB work as expected (that means: conection string ok):
Scaffold-DbContext -Connection "Provider=Microsoft.ACE.OLEDB.12.0;Jet OLEDB:Engine Type=4;Data Source=D:\Temp\Access97EF\test.mdb;" -Provider EntityFrameworkCore.Jet -OutputDir Models
Build started...
Build succeeded.

**Comment by lcrumbling** [comment]
> 
> 
> > I've always thought using Microsoft.Jet.OLEDB.4.0 for the Provider ()like the below) is more recommended. I have a Access 95 db that I use as such `Provider=Microsoft.Jet.OLEDB.4.0;Data Source=c:\\path\\db.mdb;`
> 
> FWIW, I believe 4.0 is Access 2000; OP mentions 97, which is 3.5: Provider=Microsoft.Jet.OLEDB.3.5;Data Source=c:\path\db.mdb;
> 
> That being said, I believe 4.0 will work is 97. It -may- try to convert it to the 2000 format, though.

Mmmmm, google says: Provider=Microsoft.Jet.OLEDB.3.5.1;

**Comment by ChrisJollyAU** [comment]
4.0 would be up to access 200 I believe. I know I use that version with a access 95 db

**Comment by juliokele** [comment]
> It looks like something is going wrong in the `ExtractFileNameFromConnectionString` function.
> 
> The likely candidate is a problem with your connection string. Are you sure you've set the connection string properly?
> 
> It looks like by the time it gets to performing a regex on it to find the value of the Data Source, your connection string is NULL
> 
> While it might not be part of the problem, I'm not too sure on using ACE.OLEDB.12.0 and setting the Engine Type to 4. Perhaps @lauxjpn can chime in on whether it would work
> 
> I've always thought using Microsoft.Jet.OLEDB.4.0 for the Provider ()like the below) is more recommended. I have a Access 95 db that I use as such `Provider=Microsoft.Jet.OLEDB.4.0;Data Source=c:\\path\\db.mdb;`

same error with Microsoft.Jet.OLEDB.4.0:
"Provider=Microsoft.Jet.OLEDB.4.0;Data Source=D:\\Temp\\Access97EF\\test.mdb;"

**Comment by lauxjpn** [comment]
@juliokele Please post the code that calls the `UseJet...()` method.

From the output you posted, this should be either your `ModelContext.OnConfiguring()` or `ModelContext.ModelContext()` method.

---

In a vanilla console project, adding a migration and updating the database works as expected on my end:

<details>
<summary>Program.cs</summary>

```c#
using Microsoft.EntityFrameworkCore;
using Microsoft.Extensions.Logging;

namespace IssueConsoleTemplate
{
    public class Cookie
    {
        public int CookieId { get; set; }
        public string Name { get; set; }
    }

    public class Context : DbContext
    {
        public DbSet<Cookie> Cookies { get; set; }

        protected override void OnConfiguring(DbContextOptionsBuilder optionsBuilder)
        {
            optionsBuilder
                .UseJetOleDb(@"Provider=Microsoft.ACE.OLEDB.12.0;Jet OLEDB:Engine Type=4;Data Source=E:\Sources\JetIssues\Issue106\IssueConsoleTemplate\Issue106.mdb;")
                .UseLoggerFactory(
                    LoggerFactory.Create(
                        b => b
                            .AddConsole()
                            .AddFilter(level => level >= LogLevel.Information)))
                .EnableSensitiveDataLogging()
                .EnableDetailedErrors();
        }

        protected override void OnModelCreating(ModelBuilder modelBuilder)
        {
            modelBuilder.Entity<Cookie>(
                entity =>
                {
                    entity.HasData(
                        new Cookie {CookieId = 1, Name = "Basic"},
                        new Cookie {CookieId = 2, Name = "Chocolate Chip"});
                });
        }
    }

    internal class Program
    {
        private static void Main()
        {
        }
    }
}
```

</details>

<details>
<summary>Console output</summary>

```pwsh
PS E:\Sources\JetIssues\Issue106\IssueConsoleTemplate> dotnet ef migrations add Initial
Build started...
Build succeeded.
warn: Microsoft.EntityFrameworkCore.Model.Validation[10400]
      Sensitive data logging is enabled. Log entries and exception messages may include sensitive application data, this mode should only be enabled during development.
info: Microsoft.EntityFrameworkCore.Infrastructure[10403]
      Entity Framework Core 3.1.15 initialized 'Context' using provider 'EntityFrameworkCore.Jet' with options: DataAccessProviderFactory SensitiveDataLoggingEnabled DetailedErrorsEnabled
Done. To undo this action, use 'ef migrations remove'

PS E:\Sources\JetIssues\Issue106\IssueConsoleTemplate> dotnet ef database update
Build started...
Build succeeded.
warn: Microsoft.EntityFrameworkCore.Model.Validation[10400]
      Sensitive data logging is enabled. Log entries and exception messages may include sensitive application data, this mode should only be enabled during development.
info: Microsoft.EntityFrameworkCore.Infrastructure[10403]
      Entity Framework Core 3.1.15 initialized 'Context' using provider 'EntityFrameworkCore.Jet' with options: DataAccessProviderFactory SensitiveDataLoggingEnabled DetailedErrorsEnabled
info: Microsoft.EntityFrameworkCore.Database.Command[20101]
      Executed DbCommand (903ms) [Parameters=[], CommandType='Text', CommandTimeout='60']
      CREATE DATABASE 'E:\Sources\JetIssues\Issue106\IssueConsoleTemplate\Issue106.mdb';
info: Microsoft.EntityFrameworkCore.Database.Command[20101]
      Executed DbCommand (13ms) [Parameters=[], CommandType='Text', CommandTimeout='30']
      CREATE TABLE `__EFMigrationsHistory` (
          `MigrationId` varchar(150) NOT NULL,
          `ProductVersion` varchar(32) NOT NULL,
          CONSTRAINT `PK___EFMigrationsHistory` PRIMARY KEY (`MigrationId`)
      );
info: Microsoft.EntityFrameworkCore.Database.Command[20101]
      Executed DbCommand (75ms) [Parameters=[], CommandType='Text', CommandTimeout='30']
      SELECT * FROM `INFORMATION_SCHEMA.TABLES` WHERE `TABLE_NAME` = '__EFMigrationsHistory'
info: Microsoft.EntityFrameworkCore.Database.Command[20101]
      Executed DbCommand (20ms) [Parameters=[], CommandType='Text', CommandTimeout='30']
      SELECT `MigrationId`, `ProductVersion`
      FROM `__EFMigrationsHistory`
      ORDER BY `MigrationId`;
info: Microsoft.EntityFrameworkCore.Migrations[20402]
      Applying migration '20211102100140_Initial'.
Applying migration '20211102100140_Initial'.
info: Microsoft.EntityFrameworkCore.Database.Command[20101]
      Executed DbCommand (1ms) [Parameters=[], CommandType='Text', CommandTimeout='30']
      CREATE TABLE `Cookies` (
          `CookieId` counter NOT NULL,
          `Name` longchar NULL,
          CONSTRAINT `PK_Cookies` PRIMARY KEY (`CookieId`)
      );
info: Microsoft.EntityFrameworkCore.Database.Command[20101]
      Executed DbCommand (3ms) [Parameters=[], CommandType='Text', CommandTimeout='30']
      INSERT INTO `Cookies` (`CookieId`, `Name`)
      VALUES (1, 'Basic');
info: Microsoft.EntityFrameworkCore.Database.Command[20101]
      Executed DbCommand (1ms) [Parameters=[], CommandType='Text', CommandTimeout='30']
      INSERT INTO `Cookies` (`CookieId`, `Name`)
      VALUES (2, 'Chocolate Chip');
info: Microsoft.EntityFrameworkCore.Database.Command[20101]
      Executed DbCommand (1ms) [Parameters=[], CommandType='Text', CommandTimeout='30']
      INSERT INTO `__EFMigrationsHistory` (`MigrationId`, `ProductVersion`)
      VALUES ('20211102100140_Initial', '3.1.15');
Done.
```

</details>

**Comment by juliokele** [comment]
please try without "UseLoggerFactory".
Without Logging always failed,
with Logging works.

**Comment by juliokele** [comment]
"Jet OLEDB:Engine Type=4;" parameter ignored,
database version not as expected (Access2003, Access97)

**Comment by lauxjpn** [comment]
@juliokele We fixed the issue from the OP with PR #107.
Try the latest [daily build](https://github.com/bubibubi/EntityFrameworkCore.Jet#daily-builds).

**Comment by juliokele** [comment]
only the "System.ArgumentNullException" error part fixed, this is a new issue
